src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/Monitor.java

```diff
@@ -139,42 +127,114 @@ public void stopMonitoring(MonitorConnectionContext context) {
+            }
+          }
+        }
+
+        if (!this.activeContexts.isEmpty()) {
+
           final long statusCheckStartTimeNano = this.getCurrentTimeNano();
-          this.contextLastUsedTimestampNano.set(statusCheckStartTimeNano);
+          this.contextLastUsedTimestampNano = statusCheckStartTimeNano;
 
           final ConnectionStatus status =
-              checkConnectionStatus(this.getConnectionCheckTimeoutMillis());
+              checkConnectionStatus(this.nodeCheckTimeoutMillis);
+
+          long delayMillis = -1;
+          MonitorConnectionContext monitorContext;
+          MonitorConnectionContext firstAddedMonitorContext = null;
+
+          while ((monitorContext = this.activeContexts.poll()) != null) {
+
+            synchronized (monitorContext) {
+              // If context is already invalid, just skip it
+              if (!monitorContext.isActiveContext()) {
+                continue;
+              }
+
+              if (firstAddedMonitorContext == monitorContext) {
+                // this context has already been processed by this loop
+                // add it to the queue and exit this loop
+                //System.out.println("exit loop");
+                this.activeContexts.add(monitorContext);
+                break;
+              }
 
-          for (MonitorConnectionContext monitorContext : this.contexts) {
-            monitorContext.updateConnectionStatus(
-                statusCheckStartTimeNano,
-                statusCheckStartTimeNano + status.elapsedTimeNano,
-                status.isValid);
+              // otherwise, process this context
+              monitorContext.updateConnectionStatus(
+                  this.hostInfo.getHostPortPair(),
+                  statusCheckStartTimeNano,
+                  statusCheckStartTimeNano + status.elapsedTimeNano,
+                  status.isValid);
+
+              // If context is still valid and node is still healthy, it needs to continue updating this context
+              if (monitorContext.isActiveContext() && !monitorContext.isNodeUnhealthy()) {
+                this.activeContexts.add(monitorContext);
+                if (firstAddedMonitorContext == null) {
+                  firstAddedMonitorContext = monitorContext;
+                }
+
+                if (delayMillis == -1 || delayMillis > monitorContext.getFailureDetectionIntervalMillis()) {
+                  delayMillis = monitorContext.getFailureDetectionIntervalMillis();
+                }
+              }
+            }
+          }
+
+          if (delayMillis == -1) {
+            // No active contexts
+            delayMillis = THREAD_SLEEP_WHEN_INACTIVE_MILLIS;
+          } else {
+            delayMillis -= status.elapsedTimeNano;
+            // Check for min delay between node health check
+            if (delayMillis < MIN_CONNECTION_CHECK_TIMEOUT_MILLIS) {
+              delayMillis = MIN_CONNECTION_CHECK_TIMEOUT_MILLIS;
+            }
```

Comment:
should the elapsedTime subtraction come after this reset to the MIN_CONNECTION timeout? For example, if we spent 1s on the status check and the MIN_CONNECTION timeout is 3s, we only have to wait 2s

Reply:
I was thinking about this scenario but I guess it makes sense to avoid checking node status too often. So MIN_CONNECTION is applied to delayMillis.